soroban-env-host/src/native_contract/token/allowance.rs

```diff
@@ -21,17 +28,91 @@ pub fn write_allowance(
     from: Address,
     spender: Address,
     amount: i128,
+    expiration: u32,
 ) -> Result<(), HostError> {
     let key = DataKey::Allowance(AllowanceDataKey { from, spender });
-    e.put_contract_data(
-        key.try_into_val(e)?,
-        amount.try_into_val(e)?,
-        StorageType::Persistent,
-        ().into(),
-    )?;
+
+    // validates the expiration and then returns the ledger seq
+    // The expiration can be less than ledger seq if clearing an allowance
+    let ledger_seq = e.with_ledger_info(|li| {
+        if expiration > li.sequence_number.saturating_add(li.max_entry_expiration) {
+            Err(err!(
+                e,
+                ContractError::AllowanceError,
+                "expiration is greater than max: {} > {}",
+                expiration,
+                li.max_entry_expiration
+            ))
+        } else if amount > 0 && expiration <= li.sequence_number {
+            Err(err!(
+                e,
+                ContractError::AllowanceError,
+                "expiration must be greater than ledger sequence: {} <= {}",
+                expiration,
+                li.sequence_number
+            ))
+        } else {
+            Ok(li.sequence_number)
+        }
+    })?;
+
+    if let Ok(allowance) = e.get_contract_data(key.try_into_val(e)?, StorageType::Temporary) {
+        let mut updated_allowance: AllowanceValue = allowance.try_into_val(e)?;
+
+        updated_allowance.amount = amount;
+        let old_expiration = updated_allowance.expiration_ledger;
+        updated_allowance.expiration_ledger = expiration;
+
+        e.put_contract_data(
+            key.try_into_val(e)?,
+            updated_allowance.try_into_val(e)?,
+            StorageType::Temporary,
+            ().into(),
+        )?;
+
+        if old_expiration < expiration && amount > 0 {
+            e.bump_contract_data(
+                key.try_into_val(e)?,
+                StorageType::Temporary,
+                (expiration - ledger_seq).into(),
+            )?;
+        }
+    } else if amount > 0 {
+        //New allowance
+        let val = AllowanceValue {
+            amount,
+            expiration_ledger: expiration,
+        };
+        e.put_contract_data(
```

Comment:
Can we avoid duplication and do something like `let allowance_value = if {} else if {} else {None}` (i.e. use option to handle case when no write happens).

Reply:
Also, I'm not sure if the check for `amount > 0` is needed; there is some risk of getting an incorrect footprint.

Reply:
The code gets a little confusing due to the logic around the old and new expiration. And the `amount > 0` check is there so you don't pay and bump an entry that no one needs to access.

Reply:
I'm just thinking about scenario where one sends tx for allowing N USD, then decides to revoke it, sends preflight request for 0, doesn't get allowance in the footprint and proceeds to failing the allowance amendenment (if allowance for N is externalized, that is). Not sure how realistic this is, but the general idea is that the more stable footprint is, the better. I don't feel that strongly about this particular case though.

Reply:
Does that scenario apply here? The amount > 0 is for the bump. Anyways I don't think we need to optimize for that type of scenario.